### ISSUE TITLE
Incorrect Integer Conversion Query appears to fail to capture integer parsing

### ISSUE THREAD

**Issue by tomtau** [github/codeql-go#739]
Parsing signed integers consumes one bit less than their unsigned counterparts. The usage of [strconv.ParseUint](https://golang.org/pkg/strconv/#ParseUint) to parse a signed integer out of a string returns an unsigned 64-bit integer uint64. I'd expect the query to detect, for example, this uint64 if casted with the wrong constant bitsize:
```go
    u64, err := strconv.ParseUint(str, 10, 64)
    if err != nil {
        panic(err)
    }
    i64 := int64(u64)
```
which ideally should have been:
```go
    u64, err := strconv.ParseUint(str, 10, 63)
    if err != nil {
        panic(err)
    }
    i64 := int64(u64)
```

I'm not sure whether these cases are in that query's scope. If not, it may be worth adding a query that can capture these integer parsing-related conversion bug patterns.

**Comment by smowton** [comment]
It does apply in principle, though it seems a little niche given that the user could have simply used `ParseInt(...64)` there? Is this a pattern you've seen happen often?

**Comment by tomtau** [comment]
@smowton Yes, one alternative is to use `ParseInt` and add a validation check that it's a positive number (because `ParseInt` can interpret strings with leading signs).
The pattern I posted is one common source of bugs in several Go codebases I've seen: it usually appears in the client package context where its code reads several user inputs (from CLI arguments, environments or configs) and then converts them into desired application structures. Two "whys":

1. Why do people use `ParseUint` instead of `ParseInt`? They expect positive numbers without leading signs from user inputs, so it's more convenient and less error-prone (i.e. no extra validation code that one could forget to add is needed).
2. Why do people store those numbers as signed integers if they are expected to be positive? They worry about (unchecked / default) integer arithmetic operations where e.g. the underflow behaviour from subtractions of unsigned integers could lead to subtle bugs.

Do you think this pattern should be captured by the incorrect integer conversion query?

**Comment by smowton** [comment]
I think yes in principle; this might be a little messy to implement though so it could take a little time to get to.
